File: DRP-nodejs/drp-mesh/lib/packet.js
const DRP_ErrorCode = {
    BADREQUEST: 400, // Bad Request - User provided bad input
    UNAUTHORIZED: 401, // Unauthorized - User doesn't have rights
    NOTFOUND: 404, // Not Found - Target object not found (use for PathCmd when object not found)
    SVCTIMEOUT: 408, // Request Timeout - Valid request, timed out
    SVCERR: 500, // Internal Server Error - Valid request, failure on server side
    UNAVAILABLE: 503, // Service Unavailable - Service unavailable to answer requests (use for RPC when service not found)
    GWTIMEOUT: 504, // Gateway Timeout - Broker sent command to service but did not receive response in time
    NOSTORAGE: 507, // Insufficient Storage - The server does not have storage to process request
    LOOP: 508, // Loop Detected - The server detected an infinite loop while processing the request
}

class DRP_CmdError extends Error {
    /**
     * DRP Command Error
     * @param {string} message
     * @param {number} code
     * @param {string} source
     */
    constructor(message, code, source) {
        super(message);
        this.name = "DRPCmdError";
        this.code = code
        this.source = source
    }

    toJSON() {
        return {
            name: this.name,
            code: this.code,
            message: this.message,
            source: this.source,
            stack: this.stack
        }
    }
}

class DRP_Packet {
    /**
     * 
     * @param {string} type Packet Type
     * @param {DRP_RouteOptions} routeOptions Route Options
     * @param {string} token Reply Token
     */
    constructor(type, routeOptions, token) {
        this.type = type;
        this.routeOptions = routeOptions || null;
        this.token = token;
    }
}

class DRP_Cmd extends DRP_Packet {
    /**
     * 
     * @param {string} serviceName DRP Service Name
     * @param {string} method Method Name
     * @param {Object} params Method Parameters
     * @param {number} token Reply Token
     * @param {DRP_RouteOptions} routeOptions Route Options
     * @param {string} serviceInstanceID Execute on specific Service Instance
     * @param {string} limitScope Limit execution scope (local|zone|global)
     */
    constructor(serviceName, method, params, token, routeOptions, serviceInstanceID, limitScope) {
        super("cmd", routeOptions, token);
        this.method = method;
        this.params = params;
        this.serviceName = serviceName;
        this.serviceInstanceID = serviceInstanceID;
        this.limitScope = limitScope;
    }
}

/**
 * @typedef {Object} DRP_Reply_Error
 * @property {number} code Error code
 * @property {string} msg Error message
 */

/**
 * For RPC commands, the "status" field tells whether or not the command was
 * received and processed by the target service instance.  The "err.code" contains
 * the results from the target service instance.
 */

class DRP_Reply extends DRP_Packet {
    /**
     * 
     * @param {number} token Reply Token
     * @param {number} status Status [0=failed,1=final packet,2=continue]
     * @param {DRP_Reply_Error} err Error Object
     * @param {any} payload Reply Payload
     * @param {DRP_RouteOptions} routeOptions Route Options
     */

    constructor(token, status, err, payload, routeOptions) {
        super("reply", routeOptions, token);
        this.status = status;
        this.err = err;
        this.payload = payload;
    }
}

class DRP_RouteOptions {
    /**
     * 
     * @param {string} srcNodeID Source Node ID
     * @param {string} tgtNodeID Target Node ID
     * @param {string[]} routeHistory List of Nodes used as proxies; could be used to calculate TTL
     */
    constructor(srcNodeID, tgtNodeID, routeHistory) {
        this.srcNodeID = srcNodeID;
        this.tgtNodeID = tgtNodeID;
        this.routeHistory = routeHistory || [];
    }
}

module.exports = {
    DRP_Packet: DRP_Packet,
    DRP_Cmd: DRP_Cmd,
    DRP_Reply: DRP_Reply,
    DRP_RouteOptions: DRP_RouteOptions,
    DRP_CmdError: DRP_CmdError,
    DRP_ErrorCode: DRP_ErrorCode
};

File: DRP-go/drpmesh/packet.go
package drpmesh

import (
	"encoding/json"
)

// CreateCmd returns a Cmd object
func CreateCmd(method *string, params *CmdParams, serviceName *string, token *int, serviceInstanceID *string, routeOptions *RouteOptions) *Cmd {
	drpCmd := &Cmd{}
	drpCmd.Type = "cmd"
	drpCmd.RouteOptions = routeOptions
	drpCmd.Token = token
	drpCmd.Method = method
	drpCmd.Params = params
	drpCmd.ServiceName = serviceName
	drpCmd.ServiceInstanceID = serviceInstanceID
	return drpCmd
}

// CreateReply returns a Reply object
func CreateReply(status int, payload interface{}, token *int, routeOptions *RouteOptions) *Reply {
	drpReply := &Reply{}
	drpReply.Type = "reply"
	drpReply.RouteOptions = routeOptions
	drpReply.Token = token
	drpReply.Status = status
	drpReply.Payload = payload
	return drpReply
}

// Packet describes the base attributes of a DRP packet
type Packet struct {
	Type         string        `json:"type"`
	RouteOptions *RouteOptions `json:"routeOptions"`
	Token        *int          `json:"token"`
}

// Cmd is a DRP packet sent when issuing a command
type Cmd struct {
	Packet
	Method            *string    `json:"method"`
	Params            *CmdParams `json:"params"`
	ServiceName       *string    `json:"serviceName"`
	ServiceInstanceID *string    `json:"serviceInstanceID"`
}

// CmdParams - DRP Cmd parameters
type CmdParams map[string]*json.RawMessage

// ToJSON converts the packet to a JSON byte array
func (dc *Cmd) ToJSON() []byte {
	buff, _ := json.Marshal(dc)
	return buff
}

// CmdOut is a DRP packet sent when issuing a command
type CmdOut struct {
	Packet
	Method            *string     `json:"method"`
	Params            interface{} `json:"params"`
	ServiceName       *string     `json:"serviceName"`
	ServiceInstanceID *string     `json:"serviceInstanceID"`
}

// ToJSON converts the packet to a JSON byte array
func (dc *CmdOut) ToJSON() []byte {
	buff, _ := json.Marshal(dc)
	return buff
}

// Reply is a DRP packet sent when replying to a command
type Reply struct {
	Packet
	Status  int         `json:"status"`
	Payload interface{} `json:"payload"`
}

// ToJSON converts the packet to a JSON byte array
func (dr *Reply) ToJSON() []byte {
	buff, _ := json.Marshal(dr)
	return buff
}

// ReplyIn is used to unmarshal Reply packets we get back after sending a command
type ReplyIn struct {
	Packet
	Status  int              `json:"status"`
	Payload *json.RawMessage `json:"payload"`
}

// ToJSON converts the packet to a JSON byte array
func (dri *ReplyIn) ToJSON() []byte {
	buff, _ := json.Marshal(dri)
	return buff
}

// RouteOptions is an optional Packet parameter used to take advantage of control plane routing
type RouteOptions struct {
	SrcNodeID    *string  `json:"srcNodeID"`
	TgtNodeID    *string  `json:"tgtNodeID"`
	RouteHistory []string `json:"routeHistory"`
}
